src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/WhenShouldBeUsedCheck.java

```diff
@@ -0,0 +1,176 @@
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Ensures that {@code when} is used instead of a single {@code if}
+ * statement inside a case block.
+ * </p>
+ * <p>
+ * Java 21 has introduced enhancements for switch statements and expressions
+ * that allow the use of patterns in case labels. The {@code when} keyword is used to specify
+ * condition for a case label, also called as guarded case labels. This syntax is more readable
+ * and concise than the single {@code if} statement inside the pattern match block.
+ * </p>
+ * <p>
+ * See the <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-14.html#jls-Guard">
+ * Java Language Specification</a> for more information about guarded case labels.
+ * </p>
+ * <p>
+ * See the <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-14.html#jls-14.30">
+ * Java Language Specification</a> for more information about patterns.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code when.should.be.used}
+ * </li>
+ * </ul>
+ *
+ * @since 10.18.0
+ */
+
+@FileStatefulCheck
+public class WhenShouldBeUsedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "when.should.be.used";
+
+    /**
+     *  An array of acceptable statements inside the case block.
+     */
+    private static final int[] ACCEPTABLE_STATEMENTS_INSIDE_CASE_BLOCK = {
+        TokenTypes.LITERAL_IF,
+        TokenTypes.LITERAL_BREAK,
+        TokenTypes.EMPTY_STAT,
+        TokenTypes.RCURLY,
```

Comment:
How is this a statement?

Reply:
This is not a statement actually but it is a `SLIST` child according to AST. should we change the name of this field?

Reply:
Just move it to the private method and rename it :)

Reply:
done